Library/DasharoVariablesLib/DasharoVariablesLib.c

```diff
@@ -33,37 +33,6 @@ typedef struct {
   UINT32  Attributes;  // EFI variable attributes for this variable.
 } VAR_INFO;
 
-// List of Dasharo EFI variables in gDasharoSystemFeaturesGuid namespace that
-// are created if missing.
-STATIC CHAR16 *mAutoCreatedVariables[] = {

```

Comment:
I also thought about adding the condition to the table. 

> not every variable is controlled by boolean PCD.

Some variables/options are controlled by mere exposing the menu the option is placed in. That should suffice for now.

Reply:
> Short of creating a callback for each variable's condition, I have no idea how to do this for more complex cases.

* Encode expressions in strings and write a parser to evaluate them :)  Readable, but lacks compile-time checks.
* Something like this might probably work and should catch invalid PCD names:
   ```c
   enum pcd_op {
       PCD_VAL_FIXED_PCD,
       PCD_VAL_INT,

       PCD_OP_EQ,
       PCD_OP_NE,
       PCD_OP_NOT,
   };
   struct pcd_expr {
       enum pcd_op op;
       int val;
       struct pcd_expr *lhs;
       struct pcd_expr *rhs;
   };

   #define PCD_INT(v) &(struct pcd_expr){ .op = PCD_VAL_INT, .val = (v) }
   #define PCD_FIXED(v) &(struct pcd_expr){ .op = PCD_VAL_FIXED_PCD, .val = (v) }
   #define PCD_EQ(l, r) &(struct pcd_expr){ .op = PCD_OP_EQ, .lhs = (l), .rhs = (r) }
   #define PCD_NE(l, r) &(struct pcd_expr){ .op = PCD_OP_NE, .lhs = (l), .rhs = (r) }
   #define PCD_NOT(e) &(struct pcd_expr){ .op = PCD_OP_NOT, .lhs = (e) }

   PCD_NE(PCD_FIXED(PcdDefaultPowerFailureState), PCD_INT(POWER_FAILURE_STATE_HIDDEN))

   // function to evaluate this isn't hard to write, complex expressions won't be readable
   ```

Reply:
Just did something simpler: 2ef377f5b7f52184191218ecef877c49f928b269

Reply:
This will work as long as data is constant and known at build time, which is why you had to use `FixedPcdGet*()` to access it. I'm fine with that, but please use the same `Fixed*` accessors in `STORE_VAR_IF`, and add a comment mentioning it above `mAutoCreatedVariables` for whoever will be adding more entries in the future. I'm pretty sure edk2 build system would find a way to obscure the error message if someone were to add non-fixed PCD here.

Reply:
> This will work as long as data is constant and known at build time, which is why you had to use FixedPcdGet*() to access it

And it is totally fine, because we intend to use only build-time fixed PCDs.

> I'm fine with that, but please use the same Fixed* accessors in STORE_VAR_IF, and add a comment mentioning it above mAutoCreatedVariables for whoever will be adding more entries in the future.

Got it.

Reply:
Extended comment for the array: https://github.com/Dasharo/DasharoModulePkg/pull/50/commits/da728a08d5e3da8d4aa5480cb3b3498d488fb38f#diff-84cb8f5e6369580047afd951437a6ee42307a2b38a7a66a8f49bd356770e689aR43-R45

FixedPcdGetBool: 9a6266ff5923e535cbd909d810e03a4ba21d9318